cluster_mapping.go

```diff
@@ -0,0 +1,202 @@
+type clusterQueryPlanner struct {
+	clock         clock.Clock
+	log           xlog.Logger
+	shardMappings atomic.Value
+}
+
+// newClusterQueryPlanner returns a new cluster query planner given a set of initial mappings
+func newClusterQueryPlanner(
+	initialMappings ShardClusterMappings,
+	clock clock.Clock,
+	log xlog.Logger) *clusterQueryPlanner {
+	p := &clusterQueryPlanner{
+		clock: clock,
+		log:   log,
+	}
+
+	p.updateShardMappings(initialMappings)
+	return p
+}
+
+// updateShardMappings updates the shard mappings
+func (p *clusterQueryPlanner) updateShardMappings(shardMappings ShardClusterMappings) {
+	p.shardMappings.Store(shardMappings)
+}
+
+// buildClusterQueryPlan takes a set of queries and returns the set of clusters
+// that need to be queries, along with the time range for each query.  Assumes
+// the mapping rules are sorted by cutoff time, with the most recently applied
+// rule appearing first.
+func (p *clusterQueryPlanner) buildClusterQueryPlan(shard uint32, queries []query) ([]clusterQuery, error) {
+	shardMappings := p.shardMappings.Load().(ShardClusterMappings)
+
+	cqueries := make([]clusterQuery, 0, len(queries))
+	for _, q := range queries {
+		// Find the mappings for the query retention period
+		// TODO(mmihic): and resolution?
+		mappings := shardMappings.MappingsFor(shard, q.RetentionPolicy)
+		if mappings == nil {
+			// No mappings for this retention policy, skip it
+			continue
+		}
+
+		// Find all of the rules that apply to the query time range.
+		p.log.Debugf("building query plan from %v to %v for period %s over %d mappings",
+			q.Range.Start, q.Range.End, q.RetentionPolicy.RetentionPeriod().Duration(), len(mappings))
+
+		for _, m := range mappings {
+			if !m.CutoffTime().IsZero() && m.CutoffTime().Before(q.Range.Start) {
+				// We've reached a rule that ends before the start of the query - no more rules apply
+				p.log.Debugf("cluster mapping cutoff time %s before query start %s, stopping build",
+					m.CutoffTime(), q.Range.Start)
+				break
+			}
+
+			if !m.ReadCutoverTime().IsZero() && m.ReadCutoverTime().After(q.Range.End) {
+				// We've reached a rule that doesn't apply by the time the query ends - keep going until
+				// we fnd an older rule which might apply
+				p.log.Debugf("cluster mapping cutover time %s after query end %s, skipping rule",
+					m.ReadCutoverTime(), q.Range.End)
+				continue
+			}
+
+			// Capture the cluster to query...
+			cq := clusterQuery{
+				Range: xtime.Range{
+					Start: q.Range.Start,
+					End:   q.Range.End,
+				},
+				RetentionPolicy: q.RetentionPolicy,
+				cluster:         m.Cluster(),
+			}
+
+			// ...and bound the query time by the mapping time range
+			if !m.CutoffTime().IsZero() && m.CutoffTime().Before(cq.Range.End) {
+				cq.Range.End = m.CutoffTime()
+			}
+
+			if !m.ReadCutoverTime().IsZero() && m.ReadCutoverTime().After(cq.Range.Start) {
+				cq.Range.Start = m.ReadCutoverTime()
+			}
+
+			if cq.Range.IsEmpty() {
```

Comment:
Right now `range.IsEmpty()` only returns true if `range.Start == range.End`, we might need to relax that to return true if `range.Start >= range.End` to account for queries whose range start is equal to or later than range end.